fix: raise UNREACHABLE

emc415 opened the pull request:
### What I did
fix #3070: added handling for raise UNREACHABLE

### How I did it

### How to verify it

### Commit message

fix: raise UNREACHABLE

### Description for the changelog

raise UNREACHABLE was generating CompilerPanic because of how parse_Raise calls _assert_reason, and then got passed to the block which handles assert_unreachable (IRnode.from_list(["assert_unreachable", test_expr])). This was breaking because None was being passed as a test_expr to assert_unreachable, which is not valid.
Added a new block in _assert_reason which handles raise UNREACHABLE and generates "INVALID" directly without going through the same path as assert_unreachable.

### Cute Animal Picture

![Put a link to a cute animal picture inside the parenthesis-->](https://scontent-sea1-1.xx.fbcdn.net/v/t31.18172-8/23511077_1495910010491652_7112871162007898880_o.jpg?_nc_cat=111&ccb=1-7&_nc_sid=730e14&_nc_ohc=OfRxoUOufPMAX_CUq6B&_nc_ht=scontent-sea1-1.xx&oh=00_AfBrH2ImCot6_iGFJhyPOPSNMbOxigp2uyvD0kvsNH15Vg&oe=63D3F1A7)

emc415 commented:
I changed the condition passed in _assert_reason in parse_Raise from None to 0. This changes the value of test_expr passed in _assert_reason(self, test_expr, msg) from None to 0 (False also works). None is not a valid value for test_expr, whereas 0 is. This is what is causing the error "None is not allowed as an IRnode"

Importantly, 0 used to be passed in parse_Raise when it was first implemented.
This got changed somewhere down the line.

My hypothesis is that it got changed when the block that I commented out in optimizer.py got introduced. This block raises a StaticAssertion error when 0 is passed into _assert_reason. I think that when this got introduced, None must've been added to parse_Raise to handle UNREACHABLE while avoiding 0, i.e. the optimization error. I am still combing through the history to verify this, but @charles-cooper if you have memories of doing this it would be helpful; I think it was potentially somewhere around https://github.com/vyperlang/vyper/commit/efe1dbef873af6995833d1220a3d56978dfb9de0 -- feat: more arithmetic optimizations (https://github.com/vyperlang/vyper/pull/2647)
or feat: add optimizer guardrails, refactor (#2914); and i know this recent commit fixed some other parts of 2647 that were a bit too aggressive: fix: optimizer regression (https://github.com/vyperlang/vyper/pull/2868)

Anyways, I'm still going through the history to find exactly when this change happened and why, but if you remember anything regarding reasoning let me know!

Also, trying to understand if we can just remove the commented out block in optimizer.py because I don't really think that asserts or raises need to be optimized; if a dev is using them, they are using them on purpose and probably don't need them to be optimized. Let me know what you think about this.

Thanks!

emc415 commented:
btw i found the location of the change:
allow arbitrary revert reason strings (https://github.com/vyperlang/vyper/pull/2509); 
line 215 in vyper/old_codegen/stmt.py
https://github.com/vyperlang/vyper/commit/4b9d86043fcb6db67a6eed4c41548bc6ecc7e2a4#diff-f4a4033c9a6b8afdf524b51945dde4ce1d68aac72b7778f6e249c6f2035d7541R215

Still working to understand why this happened.

emc415 commented:
Because ir/optimizer.py is correct and needs to stay as it is to catch internal compiler assert bugs:
We should add ir handling of “raise_unreachable”
This is my go at doing so
It is “inspired” by handling of assert_unreachable
It generates almost identical IR but adds a POP as described in parse_Raise in codegen/stmt.py
I am still investigating the usage and meaning of ‘0’ in IRnode.from_list([“raise_unreachable”], 0), and the reason for the added POP

charles-cooper commented:
> We should add ir handling of “raise_unreachable”

very nice. that's clean, and mirrors the two way split between `assert` (for user `assert`) and straight `revert` (for user `raise`).

i think adding a new IR instruction `raise_unreachable` is unnecessary though as `invalid` is already a valid IR instruction (since it is already defined in our list of available EVM opcodes). so you can just remove the asm generation for `raise_unreachable` and emit `invalid` where you are currently emitting `raise_unreachable`.

emc415 commented:
> > We should add ir handling of “raise_unreachable”
> 
> very nice. that's clean, and mirrors the two way split between `assert` (for user `assert`) and straight `revert` (for user `raise`).
> 
> i think adding a new IR instruction `raise_unreachable` is unnecessary though as `invalid` is already a valid IR instruction (since it is already defined in our list of available EVM opcodes). so you can just remove the asm generation for `raise_unreachable` and emit `invalid` where you are currently emitting `raise_unreachable`.

Okay, thanks!!

emc415 commented:
last commit: pushed change to codegen/stmt.py to emit "invalid" instead of "raise_unreachable" so that test suite can run here
TODO: investigate whether the diff breaks anything in ASM of: --no-optimize assert False, UNREACHABLE vs. --no-optimize raise UNREACHABLE:
JUMPDEST [-PUSH1 0x00 _sym_reachable5 JUMPI-] INVALID [-_sym_reachable5 JUMPDEST _sym_external_unreachable____cleanup JUMP-]

charles-cooper commented on vyper/codegen/stmt.py:
```diff
@@ -205,24 +207,23 @@ def _get_last(ir):
     def parse_Raise(self):
         if self.stmt.exc:
+            if isinstance(self.stmt.exc, vy_ast.Name) and self.stmt.exc.id == "UNREACHABLE":
+                self.stmt.exc.id = "raise UNREACHABLE"
```
why modify the ast here?

emc415 replied:
as a way to tag it as raise unreachable so that it doesn't get passed into IRnode.from_list(["assert_unreachable", test_expr]); otherwise by the time you get to _assert_reason there's no way that I know of to distinguish it

charles-cooper replied:
that's why the test expr is passed in as None, so `assert_reason` can distinguish -- see the final branch in the `_assert_reason()` implementation.

charles-cooper commented on vyper/codegen/stmt.py:
```diff
@@ -160,11 +160,12 @@ def parse_Call(self):
-            )
-        elif isinstance(msg, vy_ast.Name) and msg.id == "raise UNREACHABLE":
-            return IRnode.from_list(["invalid"], error_msg="raise unreachable")
+            if test_expr is not None:
+                return IRnode.from_list(
+                    ["assert_unreachable", test_expr], error_msg="assert unreachable"
+                )
+            else:
+                return IRnode.from_list(["invalid"], error_msg="raise unreachable")
```
nice -- since this check is performed twice, and it's actually not entirely clear what the check is for, i would factor out the expression (`test_expr is None`) into a variable called `is_raise` and add a comment about what it is doing.

emc415 replied:
Cool! Is it possible that None can be passed into parse_Assert from an internal compiler assert? I am just thinking that if so, it may be a little misleading if it got called 'is_raise', but also it shouldn't affect behavior.

charles-cooper replied:
no, `parse_Assert` only handles user asserts -- but if you are paranoid (which is not a bad thing!) you can add a sanity check in `parse_Assert` on the test expr before passing down into `assert_reason()`.

charles-cooper commented:
this is looking good. i think we want to add a test for `raise UNREACHABLE`, you could add it to `test_assert_unreachable.py`.

emc415 commented:
cool, thanks! implemented your suggestions.

Also, I removed a line from test_basic_call_unreachable which I am pretty certain was a duplicate and not necessary. Just want to make sure to point it out to y'all in case it was indeed necessary.

charles-cooper commented:
thanks, @trocher what do you think about this fix?

trocher commented:
This is looking good to me!

charles-cooper commented on vyper/codegen/stmt.py:
```diff
@@ -205,16 +211,16 @@ def _get_last(ir):
 
     def parse_Assert(self):
         test_expr = Expr.parse_value_expr(self.stmt.test, self.context)
+        # sanity check
+        assert test_expr is not None
```
note: the Expr constructor already checks for this
